### PR TITLE
Added Documentation Style Guide; Fixes #3089

### LINKED ISSUE #3089: Style Guide (Documentation Style) [RFC]
Style Guide (Documentation)
===========================

General Guidelines
------------------

* Authors should document anything that is exported.
* Within a documentation for a procedure, a period (`.`) should follow each sentence in a comment block if there is more than one sentence in that block; otherwise, no period should be used. In other sections, complete sentences should have periods. Sentence fragments should have none.
* Documentation is parsed as RST (RestructuredText).
* Inline code should be surrounded by double tick marks (``` `` ```). With ReStructuredTest (RST), if you would like a character to immediately follow inline code (e.g., "``int8``s are great!"), escape the following character with a backslash (`\`). The preceding is typed as ``` ``int8``\s are great!```.

Module-level documentation
--------------------------

Documentation of a module is placed at the top of the module itself. Each line of documentation begins with double hashes (`##`).
Code samples are encouraged, and should follow the general RST syntax:

```nim
## The ``universe`` module computes the answer to life, the universe, and everything.
##
## .. code-block:: nim
##  echo computeAnswerString*() # "42"
```

Within this top-level comment, you can indicate the authorship and copyright of the code, which will be featured in the produced documentation.

```nim
## This is the best module ever. It provides answers to everything!
##
## :Author: Steve McQueen
## :Copyright: 1965
##
```

Users are encouraged to leave a space between the last line of top-level documentation and the beginning of Nim code (the imports, etc.).

Procs, Templates, Macros, Converters, and Iterators
---------------------------------------------------

The documentation of a procedure should begin with a capital letter and should be in present tense. Variables referenced in the documentation should be surrounded by double tick marks (``` `` ```).

```nim
    proc example1*(x: int) =
        ## Prints the value of ``x``
        echo x
```

Whenever an example of usage would be helpful to the user, please include a sample within the documentation in RST format as below.

```nim
    proc addThree*(x, y, z: int8): int =
        ## Adds three ``int8`` values, treating them as unsigned and
        ## truncating the result
        ##
        ## .. code-block:: nim
        ##  echo addThree(3, 125, 6) # -122
        result = x +% y +% z
```

The commands ``nim doc`` and ``nim doc2`` will then correctly syntax highlight the Nim code within the documentation.

Types
-----

Types should also be documented. This documentation can also contain code samples, but those are likely better placed with the functions to which they refer.

```nim
type
  NamedQueue*[T] = object ## Provides a linked data structure with names
                          ## throughout. Named for convenience. I'm making
                          ## this comment long to show how you can, too.
    name*: string ## The name of the item
    val*: T ## Its value
    next*: ref NamedQueue[T] ## The next item in the queue
```

You have some flexibility when placing the documentation:
```nim
type
  NamedQueue*[T] = object
    ## Provides a linked data structure with names
    ## throughout. Named for convenience. I'm making
    ## this comment long to show how you can, too.
    name*: string ## The name of the item
    val*: T ## Its value
    next*: ref NamedQueue[T] ## The next item in the queue
```

Make sure to place the documentation beside or within the object.

```nim
type
  ## This documentation disappears. It's basically annotating the
  ## ``type`` above.
  NamedQueue*[T] = object
    name*: string ## This becomes the main documentation for the object, which
                  ## is not what we want.
    val*: T ## Its value
    next*: ref NamedQueue[T] ## The next item in the queue

```

Var, Let, and Const
-------------------

When declaring module-wide constants and values, documentation is encouraged. The placement of doc comments is similar to the ``type`` sections.

```nim
const
  X* = 42 ## An awesome number
  SpreadArray* = [
    [1,2,3],
    [2,3,1],
    [3,1,2],
  ] ## Doc comment for ``SpreadArray``
```

Placement of comments in other areas is usually allowed, but will not become part of the documentation output.

```nim
const
  BadMathVals* = [
    3.14, ## pi
    2.72, ## e
    0.58, ## gamma
  ] ## A bunch of badly rounded values
```

In the produced documentation, the comments beside the elements of ``BadMathVals`` do not make it into the final documentation.

As a final note, Nim supports Unicode in comments just fine, so the above can be replaced with the following:

```nim
const
  BadMathVals* = [
    3.14, ## π
    2.72, ## e
    0.58, ## γ
  ] ## A bunch of badly rounded values (including π!)
```

### PR DESCRIPTION
**Pull request by apense**
Added link to it from contributing.rst

### CODE REVIEW

**Comment by ReneSac** [comment]
I think that you should add references to it in more places. See the recently merged nep1: https://github.com/nim-lang/Nim/pull/3098/files

**Comment by apense** [comment]
@ReneSac I think the documentation part of contributing.rst is from an old PR I made, so I need to update that anyways. Should contributing.rst link to nep1 too? Or they should separately link to the styleguide? Contributing.rst just shows up on GitHub, right? And nep1 will actually be in the HTML docs?

**Comment by ReneSac** [comment]
I think at least NEP1 and this documentation style guide should link to each other and both be in the HTML docs.  It is basically an extension to NEP1. And calling it simply styleguide is confusing, as NEP1 is also called style guide:
```
+    - | `Nim Style Guide <nep1.html>`_
+      | The stylistic conventions that Nim's official projects adhere to.
```

And it may be a little late, but what is the rationale behind forbidding periods for single sentences? I would prefer periods in all documentation sentences.

**Comment by dom96** [comment]
There appear to be many errors in your rst. https://github.com/apense/Nim/blob/styledoc/styleguide.rst

**Comment by apense** [comment]
Oh this was written in Markdown. I forgot. Is there a way to keep it in Markdown or do I need to convert to RST?

**Comment by Araq** [comment]
> Oh this was written in Markdown. I forgot. Is there a way to keep it in Markdown or do I need to convert to RST?

markdown is against our guidelines. ;-)

**Comment by apense** [comment]
@Araq I'll change it, then :-D.
@ReneSac The logic went like this: it used to be in lowercase with a period, so that the proc and documentation was read as a sentence (``thisProc() gives a good answer.``). But since the documentation doesn't appear on the same line, it doesn't really read like that, so a capital letter begins. If there's a period at the end of a capitalized sentence fragment, it might look like bad English. The more clauses you add, the less weird it looks, so when there's 2+ sentences, we should use periods. 

That being said, if you have community support for periods (always), I can change it before it's pulled. It's not too late.

I'll also be changing the file name and title so it doesn't clash with the other style guide. I guess I can do this within the same PR.

File: docstyle.rst [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+Documentation of a module is placed at the top of the module itself. Each line of documentation begins with double hashes (``##``).
+Code samples are encouraged, and should follow the general RST syntax:
+
+.. code-block:: Nim
+
+  ## The ``universe`` module computes the answer to life, the universe, and everything.
+  ##
+  ## .. code-block:: Nim
+  ##  echo computeAnswerString*() # "42"
```
Comment (Araq) [inline docstyle.rst]: Wrong *.

**Comment by ReneSac** [comment]
Well, I'm not a native english speaker. For me proper sentences, especially started with a capital letter, should end with a period, and sentences w/o it reminds me of instant messaging/IRC (though those don't start with a capital letter). Also, it is more consistent to have all sentences in a page ending with a period, instead of some with and some w/o.

For me it looks better like this, even for one sentence:
```
This is the same as the type unsigned long in C.
```
Than this:

```
This is the same as the type unsigned long in C
```

**Comment by apense** [comment]
@ReneSac I understand, but when there's only one clause, it tends to be dependent and looks a little weird with a period.

```nim
proc sayHello*(): string =
  ## Says hello
proc giveFivePi*(): float =
  ## Computes the value of pi to five decimal places
```
```nim
proc sayHello*(): string =
  ## Says hello.
proc giveFivePi*(): float =
  ## Computes the value of pi to five decimal places.
```

I think either is OK, but it seems like the discussion at #3089 didn't prefer a period at the end when the sentence is imcomplete (only should happen if the entire documentation is a sentence fragment). Maybe I'm imagining that. 
Not having a period after a beginning fragment seems to make a compromise between forcing full sentences and allowing the proc name to be part of the documentation.

**Comment by ReneSac** [comment]
Well, I tried to look at what some other languages are doing with their documentation style. The four I looked at use periods to end every sentence in their documentation. I couldn't find any language that uses the style proposed here in my brief search.

Swift:
https://developer.apple.com/library/prerelease/mac/documentation/General/Reference/SwiftStandardLibraryReference/index.html

Ocalm:
http://caml.inria.fr/pub/docs/manual-ocaml/libref/String.html

Rust:
https://doc.rust-lang.org/std/string/struct.DerefString.html

Scala:
http://www.scala-lang.org/api/2.10.3/index.html#scala.math.Ordering$$String$

**Comment by apense** [comment]
@ReneSac I agree. OCaml and Swift styles don't really apply here because of their sentence structure, but Rust and Scala use a period even with a sentence fragment. Plus, it makes it easier to direct. I'm gonna change that rule. Python also puts a period, but in a command style like OCaml and Swift.

I think Rust and Scala have a good thing going. I was wrong before, too. #3089 doesn't address the period thing, I think. I believe it was in some other issue where it was discussed.

Point being, ease and consistency will rule out. So I'll change the rule to "period always."

**Comment by ReneSac** [comment]
I just updated the swift link to point to a sentence fragment. I usually don't use the period for in-code comments, but for documentation comments, that will be rendered in html, I think sentences w/o periods looks weird.

**Comment by ReneSac** [comment]
What about the documentation for constants, types, object fields, etc? I think they probably also should end in periods, not only proc descriptions...

**Comment by apense** [comment]
I changed the period guidelines. What's the preference for types? 
```nim
type X* = object ## My object.
  name*: string ## Its name.
  value*: float ## The value it takes.
```
```nim
type X* = object ## My object
  name*: string ## Its name
  value*: float ## The value it takes
```

There are different ways of doing this. Suggestions? I think I left it open-ended before, then I put periods everywhere to mesh with the procs. Periods everywhere? No requirement?

**Comment by apense** [comment]
I need to change ``contributing.rst``, but to avoid a flurry of commits, I just want to make sure I have the "all clear" on the style beforehand.

**Comment by ReneSac** [comment]
Currently the type documentation is rendered in html, but the object fields documentation is printed as code... So I would be ok with optional periods for the object fields. But I guess periods for everyone is ok too for consistency.

Kinda off-topic, but I think this too should be linked from somewhere (and maybe converted to .rst):
https://github.com/nim-lang/Nim/blob/d270c8ebe8b35f4c9bdf23e4153f299e7919bdb7/doc/apis.txt

**Comment by Araq** [comment]
Are you guys aware of http://nim-lang.org/docs/docgen.html ?

**Comment by zachaysan** [comment]
I agree with ReneSac that sentences that do not end with periods look strange and I'd support their use wherever possible.

**Comment by Varriount** [comment]
I'm also with renesac. English grammer dictates that sentences almost always end with an period.

**Comment by apense** [comment]
@Araq I remember those documents, but I think that's because I've read through most of the documentation at some point (out of curiousity and to learn). 
I'm trying to figure out where all the linkages to this doc should go. I think that's the last thing to do here, now that the period issue is over. But really, links can always come later after we have a minimal set.

**Comment by dom96** [comment]
Progress on fixing rst errors?

**Comment by apense** [comment]
Sorry about the delay! I thought I had fixed all the errors in the rst. (changing ```` ```nim ```` to `` .. code-block:: nim ``, etc.). Do I check that with `docgen`?

File: docstyle.rst [hunk 2]
```diff
@@ -0,0 +1,140 @@
+Documentation Style
+===================
+
+General Guidelines
+------------------
+
+* Authors should document anything that is exported.
+* Within documentation, a period (`.`) should follow each sentence (or sentence fragment) in a comment block. The documentation may be limited to one sentence fragment, but if multiple sentences are within the documentation, each sentence after the first should be complete and in present tense.
+* Documentation is parsed as ReStructuredText (RST).
+* Inline code should be surrounded by double tick marks ("``` `` ```"). If you would like a character to immediately follow inline code (e.g., "``int8``s are great!"), escape the following character with a backslash (``\``). The preceding is typed as ``` ``int8``\s are great!```.
```
Comment (kirbyfan64) [inline docstyle.rst]: The double tick marks are not showing up correctly. Try:

    ``````

(six tick marks)

File: docstyle.rst [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  ## :Author: Steve McQueen
+  ## :Copyright: 1965
+  ##
+
+Leave a space between the last line of top-level documentation and the beginning of Nim code (the imports, etc.).
+
+Procs, Templates, Macros, Converters, and Iterators
+---------------------------------------------------
+
+The documentation of a procedure should begin with a capital letter and should be in present tense. Variables referenced in the documentation should be surrounded by double tick marks (``` `` ```).
```
Comment (kirbyfan64) [inline docstyle.rst]: Same thing here.

**Comment by Araq** [comment]
Use ``nim rst2hmtl  foo.rst`` to check it, but Nim's Rst parser is less picky about newlines in general.

**Pull request merged**